### ISSUE TITLE
Iceberg REST Catalog might not be retrying correctly when conflict

### ISSUE THREAD

**Issue by aletc1** [trinodb/trino#15586]
When two or more INSERTs run in parallel, the first succeeds and the second fails which is OK, but Trino tries to retry the operation (because the REST Catalog returns 409) but it is sending assert-ref-snapshot-id with the previous snapshot-id and not the updated/current ref snapshot-id.
```
insert into iceberg.main.my_table values (1, cast('2020-01-01' as date), 1.0)
```
This would be the flow:
1. First commit creates snapshot-id = 2 (let's simplify numbering for sake of example and assume that table had previously current-snapshot-id = 1)
2. Second commit fails because assert-ref-snapshot-id has snapshot-id = 1 which is correct.
3. Trino reads again the table metadata from the REST catalog
4. Trino retries the commit operation but keeps assert-ref-snapshot-id with snapshot-id = 1. Should'nt this be snapshot-id = 2? (since Trino reads the updated metadata and have the updated ref with the new snapshot-id).

From Iceberg Spec:
> When two commits happen at the same time and are based on the same version, only one commit will succeed. In most cases, the failed commit can be applied to the new current version of table metadata and retried. (...) Append operations have no requirements and can always be applied.

**Comment by electrum** [comment]
cc @danielcweeks @rdblue

**Comment by rdblue** [comment]
Sounds like we need to find out how the operation could retry without refreshing the current table metadata. @aletc1 is it possible that the table load response is cached somewhere in your REST catalog?

**Comment by aletc1** [comment]
Hi @rdblue, no, I checked that the REST catalog returns fresh data (no cache). Actually, it is correct that Trino requests the metadata from the REST catalog again (because there is a 409 conflict, so you actually need to get updated metadata from the catalog). The issue is that Trino does not update the snapshot-id when retry and keeps the assertion with old value. The only way to retry a commit (if two parallel inserts occurs) is that the second one (which correctly fails) gets the latest snapshot-id from the catalog and update the assertion. 

From [Iceberg Spec](https://iceberg.apache.org/spec/):

> If the snapshot on which an update is based is no longer current, the writer must retry the update based on the new current version. Some operations support retry by re-applying metadata changes and committing, under well-defined conditions.

In case of insert it is safe to just take the new snapshot-id. For update/delete extra work is needed.

**Comment by rdblue** [comment]
I just double checked this commit path in Iceberg and I don't see what would be different in Trino. Do you see the same behavior committing in other engines?

Retries are handled internally to Iceberg and there's no difference in the retry code between REST and other catalogs. The current snapshot ID requirement is created each commit attempt based on the base metadata, which has been refreshed. That narrows the possible problems here.

Can you share more information? What was the error message and stack trace on the client? Do you have any logs from the server about the requests?

**Comment by aletc1** [comment]
Hi @rdblue ,

I cannot say if this is an iceberg library issue or trino's issue since I am looking it from the catalog side. I can share the HTTP communication trace between trino and the iceberg REST catalog API. I have no cache in between:

1. Metadata output from the first INSERT (You can see, 1 row inserted and current snapshot id is 4917725363952178803)
```json
{
   "metadata-location":"s3a://demo/lakeflow/test_concurrency_zzz",
   "metadata":{
      "last-sequence-number":2,
      "last-updated-ms":1674047312481,
      "table-uuid":"593912c5-fd82-4931-b686-9197317ee590",
      "format-version":2,
      "last-column-id":3,
      "current-schema-id":0,
      "last-partition-id":0,
      "default-spec-id":0,
      "default-sort-order-id":0,
      "current-snapshot-id":4917725363952178803,
      "location":"s3a://demo/lakeflow/test_concurrency_zzz",
      "schemas":[
         {
            "type":"struct",
            "schema-id":0,
            "fields":[
               {
                  "id":1,
                  "name":"c1",
                  "required":false,
                  "type":"int"
               },
               {
                  "id":2,
                  "name":"c2",
                  "required":false,
                  "type":"date"
               },
               {
                  "id":3,
                  "name":"c3",
                  "required":false,
                  "type":"double"
               }
            ]
         }
      ],
      "partition-specs":[
         {
            "spec-id":0,
            "fields":[
               
            ]
         }
      ],
      "sort-orders":[
         {
            "order-id":0,
            "fields":[
               
            ]
         }
      ],
      "properties":{
         "format-version":"2",
         "write.format.default":"ORC"
      },
      "snapshots":[
         {
            "sequence-number":1,
            "snapshot-id":4708628712224679730,
            "timestamp-ms":1674047297026,
            "summary":{
               "operation":"append",
               "changed-partition-count":"0",
               "total-records":"0",
               "total-files-size":"0",
               "total-data-files":"0",
               "total-delete-files":"0",
               "total-position-deletes":"0",
               "total-equality-deletes":"0"
            },
            "manifest-list":"s3a://demo/lakeflow/test_concurrency_zzz/metadata/snap-4708628712224679730-1-fd11d1c8-a6fe-41c8-bf6e-0aea81a1d83e.avro",
            "schema-id":0
         },
         {
            "sequence-number":2,
            "snapshot-id":4917725363952178803,
            "parent-snapshot-id":4708628712224679730,
            "timestamp-ms":1674047312148,
            "summary":{
               "operation":"append",
               "added-data-files":"1",
               "added-records":"1",
               "added-files-size":"514",
               "changed-partition-count":"1",
               "total-records":"1",
               "total-files-size":"514",
               "total-data-files":"1",
               "total-delete-files":"0",
               "total-position-deletes":"0",
               "total-equality-deletes":"0"
            },
            "manifest-list":"s3a://demo/lakeflow/test_concurrency_zzz/metadata/snap-4917725363952178803-1-2a50b3ae-aee3-4f5d-9261-edbcb8239593.avro",
            "schema-id":0
         }
      ],
      "refs":{
         "main":{
            "type":"branch",
            "snapshot-id":4917725363952178803
         }
      },
      "snapshot-log":[
         {
            "snapshot-id":4708628712224679730,
            "timestamp-ms":1674047297125
         },
         {
            "snapshot-id":4917725363952178803,
            "timestamp-ms":1674047312481
         }
      ]
   },
   "config":{
      
   }
}
```
2. POST http://192.168.1.177:8181/v1/namespaces/main/tables/test_concurrency_zzz. As you can see, this must fail because is expecting snapshot id 4708628712224679730 which is previous (table created) to the last INSERT. So far so good. (Remember the test is to run two INSERT in parallel. This would be the second one. It is correct that both INSERTs use 4708628712224679730 as the parent snapshot, hence this one must fail)
```json
{
   "requirements":[
      {
         "type":"assert-table-uuid",
         "uuid":"593912c5-fd82-4931-b686-9197317ee590"
      },
      {
         "type":"assert-ref-snapshot-id",
         "ref":"main",
         "snapshot-id":4708628712224679730
      }
   ],
   "updates":[
      {
         "action":"add-snapshot",
         "snapshot":{
            "sequence-number":2,
            "snapshot-id":846500503315433384,
            "parent-snapshot-id":4708628712224679730,
            "timestamp-ms":1674047312165,
            "summary":{
               "operation":"append",
               "added-data-files":"1",
               "added-records":"1",
               "added-files-size":"514",
               "changed-partition-count":"1",
               "total-records":"1",
               "total-files-size":"514",
               "total-data-files":"1",
               "total-delete-files":"0",
               "total-position-deletes":"0",
               "total-equality-deletes":"0"
            },
            "manifest-list":"s3a://demo/lakeflow/test_concurrency_zzz/metadata/snap-846500503315433384-1-24a8436d-3ffc-44f7-909b-41d0ad16a0d5.avro",
            "schema-id":0
         }
      },
      {
         "action":"set-snapshot-ref",
         "ref-name":"main",
         "snapshot-id":846500503315433384,
         "type":"branch"
      }
   ]
}

409 RESPONSE
{"type":"CommitFailedException","code":409,"description":"Snapshot-id in ref 'main' does not match"}
```
3. GET http://192.168.1.177:8181/v1/namespaces/main/tables/test_concurrency_zzz Trino actually requests the latest metadata and as you can see it is the latest one (current snapshot id = 4917725363952178803)

```json
200 RESPONSE
{
   "metadata-location":"s3a://demo/lakeflow/test_concurrency_zzz",
   "metadata":{
      "last-sequence-number":2,
      "last-updated-ms":1674047312481,
      "table-uuid":"593912c5-fd82-4931-b686-9197317ee590",
      "format-version":2,
      "last-column-id":3,
      "current-schema-id":0,
      "last-partition-id":0,
      "default-spec-id":0,
      "default-sort-order-id":0,
      "current-snapshot-id":4917725363952178803,
      "location":"s3a://demo/lakeflow/test_concurrency_zzz",
      "schemas":[
         {
            "type":"struct",
            "schema-id":0,
            "fields":[
               {
                  "id":1,
                  "name":"c1",
                  "required":false,
                  "type":"int"
               },
               {
                  "id":2,
                  "name":"c2",
                  "required":false,
                  "type":"date"
               },
               {
                  "id":3,
                  "name":"c3",
                  "required":false,
                  "type":"double"
               }
            ]
         }
      ],
      "partition-specs":[
         {
            "spec-id":0,
            "fields":[
               
            ]
         }
      ],
      "sort-orders":[
         {
            "order-id":0,
            "fields":[
               
            ]
         }
      ],
      "properties":{
         "format-version":"2",
         "write.format.default":"ORC"
      },
      "snapshots":[
         {
            "sequence-number":1,
            "snapshot-id":4708628712224679730,
            "timestamp-ms":1674047297026,
            "summary":{
               "operation":"append",
               "changed-partition-count":"0",
               "total-records":"0",
               "total-files-size":"0",
               "total-data-files":"0",
               "total-delete-files":"0",
               "total-position-deletes":"0",
               "total-equality-deletes":"0"
            },
            "manifest-list":"s3a://demo/lakeflow/test_concurrency_zzz/metadata/snap-4708628712224679730-1-fd11d1c8-a6fe-41c8-bf6e-0aea81a1d83e.avro",
            "schema-id":0
         },
         {
            "sequence-number":2,
            "snapshot-id":4917725363952178803,
            "parent-snapshot-id":4708628712224679730,
            "timestamp-ms":1674047312148,
            "summary":{
               "operation":"append",
               "added-data-files":"1",
               "added-records":"1",
               "added-files-size":"514",
               "changed-partition-count":"1",
               "total-records":"1",
               "total-files-size":"514",
               "total-data-files":"1",
               "total-delete-files":"0",
               "total-position-deletes":"0",
               "total-equality-deletes":"0"
            },
            "manifest-list":"s3a://demo/lakeflow/test_concurrency_zzz/metadata/snap-4917725363952178803-1-2a50b3ae-aee3-4f5d-9261-edbcb8239593.avro",
            "schema-id":0
         }
      ],
      "refs":{
         "main":{
            "type":"branch",
            "snapshot-id":4917725363952178803
         }
      },
      "snapshot-log":[
         {
            "snapshot-id":4708628712224679730,
            "timestamp-ms":1674047297125
         },
         {
            "snapshot-id":4917725363952178803,
            "timestamp-ms":1674047312481
         }
      ]
   },
   "config":{
      
   }
}
```

4. POST http://192.168.1.177:8181/v1/namespaces/main/tables/test_concurrency_zzz. Here comes the problem. Why assert-ref-snapshot-id keeps 4708628712224679730 since the metadata was refreshed in the previous step?
```json
{
   "requirements":[
      {
         "type":"assert-table-uuid",
         "uuid":"593912c5-fd82-4931-b686-9197317ee590"
      },
      {
         "type":"assert-ref-snapshot-id",
         "ref":"main",
         "snapshot-id":4708628712224679730
      }
   ],
   "updates":[
      {
         "action":"add-snapshot",
         "snapshot":{
            "sequence-number":2,
            "snapshot-id":846500503315433384,
            "parent-snapshot-id":4708628712224679730,
            "timestamp-ms":1674047312165,
            "summary":{
               "operation":"append",
               "added-data-files":"1",
               "added-records":"1",
               "added-files-size":"514",
               "changed-partition-count":"1",
               "total-records":"1",
               "total-files-size":"514",
               "total-data-files":"1",
               "total-delete-files":"0",
               "total-position-deletes":"0",
               "total-equality-deletes":"0"
            },
            "manifest-list":"s3a://demo/lakeflow/test_concurrency_zzz/metadata/snap-846500503315433384-1-24a8436d-3ffc-44f7-909b-41d0ad16a0d5.avro",
            "schema-id":0
         }
      },
      {
         "action":"set-snapshot-ref",
         "ref-name":"main",
         "snapshot-id":846500503315433384,
         "type":"branch"
      }
   ]
}

409 RESPONSE
{"type":"CommitFailedException","code":409,"description":"Snapshot-id in ref 'main' does not match"}
```
If you think this is not related to trino but to iceberg library, let me know to open a new issue in the propper repository. Thanks.

**Comment by aletc1** [comment]
Found the problem and it is not trino related. Actually it is arguably a problem with the Iceberg REST API client implementation which is very sensitive to the JSON payload format (e.g the http 409 respose code is not enough, it needs a payload with an error message or the message serializer will throw an exception).

```
java.lang.IllegalArgumentException: Cannot parse missing field: error
	at org.apache.iceberg.relocated.com.google.common.base.Preconditions.checkArgument(Preconditions.java:145)
	at org.apache.iceberg.rest.responses.ErrorResponseParser.fromJson(ErrorResponseParser.java:84)
	at org.apache.iceberg.rest.responses.ErrorResponseParser.fromJson(ErrorResponseParser.java:73)
	at org.apache.iceberg.rest.ErrorHandlers$DefaultErrorHandler.parseResponse(ErrorHandlers.java:141)
	at org.apache.iceberg.rest.HTTPClient.throwFailure(HTTPClient.java:120)
	at org.apache.iceberg.rest.HTTPClient.execute(HTTPClient.java:224)
	at org.apache.iceberg.rest.HTTPClient.get(HTTPClient.java:259)
	at org.apache.iceberg.rest.RESTClient.get(RESTClient.java:96)
	at org.apache.iceberg.rest.RESTClient.get(RESTClient.java:79)
	at org.apache.iceberg.rest.RESTSessionCatalog.loadInternal(RESTSessionCatalog.java:256)
	at org.apache.iceberg.rest.RESTSessionCatalog.loadTable(RESTSessionCatalog.java:271)
	at io.trino.plugin.iceberg.catalog.rest.TrinoRestCatalog.lambda$loadTable$3(TrinoRestCatalog.java:240)
	at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1708)
	at io.trino.plugin.iceberg.catalog.rest.TrinoRestCatalog.loadTable(TrinoRestCatalog.java:237)
	at io.trino.plugin.iceberg.IcebergMetadata.getTableHandle(IcebergMetadata.java:353)
	at io.trino.plugin.iceberg.IcebergMetadata.getTableHandle(IcebergMetadata.java:261)
	at io.trino.plugin.base.classloader.ClassLoaderSafeConnectorMetadata.getTableHandle(ClassLoaderSafeConnectorMetadata.java:1103)
	at io.trino.metadata.MetadataManager.lambda$getTableHandle$4(MetadataManager.java:281)
	at java.base/java.util.Optional.flatMap(Optional.java:289)
	at io.trino.metadata.MetadataManager.getTableHandle(MetadataManager.java:275)
	at io.trino.metadata.MetadataManager.getTableHandle(MetadataManager.java:262)
	at io.trino.execution.CreateTableTask.internalExecute(CreateTableTask.java:134)
	at io.trino.execution.CreateTableTask.execute(CreateTableTask.java:122)
	at io.trino.execution.CreateTableTask.execute(CreateTableTask.java:88)
	at io.trino.execution.DataDefinitionExecution.start(DataDefinitionExecution.java:145)
	at io.trino.execution.SqlQueryManager.createQuery(SqlQueryManager.java:249)
	at io.trino.dispatcher.LocalDispatchQuery.lambda$startExecution$7(LocalDispatchQuery.java:143)
	at io.trino.$gen.Trino_405____20230123_192807_2.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

Once I fixed the JSON payload, then Trino is working properly handling the correct ```org.apache.iceberg.exceptions.CommitFailedException```.  

This issue can be closed; sorry for the inconvenience.

**Comment by aletc1** [comment]
I found the issue. Trino it is working properly. As you can see in the previous http traffic log, I was returning a bad "metadata-location" in the REST API response, so I guess that Trino or Iceberg caches the metadata somehow using that field and that's why it wasnt able to get the updated metadata (because all the updates came with the same metadata-location). Once I delivered the propper metadata-location (the full .json file path), everything started to work fine.

You can close this issue, thanks for your support.

**Comment by rdblue** [comment]
Great, thanks for letting us know you solved it, @aletc1!

The metadata location is used to determine whether metadata has changed at all. If it is the same, then we don't bother to parse the metadata from the REST response.
